### PR TITLE
GDScript: statement form of "preload" and "load"

### PR DESCRIPTION
**Pull request by ghost**
This topic has been discussed before in https://github.com/godotengine/godot/issues/5570 and https://github.com/godotengine/godot/issues/12212, which is why this text is lengthy to argue this PR's cause.

One established way to load scripts and resources in GDScript seems to be `var my_scene = preload("my_scene.tscn")` for scenes and `const MyClass = preload("MyClass.gd")` for scripts. Here's a synopsis of the muddle this is bound to lead to in many cases IMHO:

```
var Loader = preload("MyClasses/A.gd") # A != Loader; as a var, should be "class_a"?; should really have been "const"
var scene = load("X/bootload.tscn") # scene != bootload
const MY_CONST = 1
var my_position = Vector2()
var U = preload("MyClasses/Utilities.gd") # U != Utilities; as a var, should be "u"?; should really have been "const"
var OtherScene = preload("Y/my_other_scene.tscn") # OtherScene != my_other_scene; "OtherScene" shows up as camel-case var, should really have been a "const"
var state = 0
```

To summarize:

* Having to manually name each loaded resource will often have programmers choose inconsistent, wrong and messy alias names for things that already do have a clear name (the resource file name)
* People get confused whether they should use `var` or `const`, `load` or `preload`, and which combination of the four
* Having `load`s and `preload`s only show up as part of `var` and `const` declarations does not help things

Even if you keep your syntax very clean, you cannot help ending up with this redundant mess:

```
const A = preload("MyClasses/A.gd")
const Utilities = preload("MyClasses/Utilities.gd")

var bootload = load("X/bootload.tscn")
var other_scene = preload("Y/my_other_scene.tscn") # should be const?

const MY_CONST = 1

var my_position = Vector2()
var state = 0
```

This PR adds `preload` and `load` as regular statements (in addition to their current role as functions) that can be used like this:

```
preload "MyClasses/A.gd"
preload "MyClasses/Utilities.gd"

preload "Y/my_other_scene.tscn"
load "X/bootload.tscn"

var my_position = Vector2()
var state = 0
```

`preload` and `load` declare suitably named constants and variables derived from the loaded resource's file name (i.e. `A`, `Utilities`, `my_other_scene` and `my_scene`); this step could contain an automatic snake-case/camel-case conversion if needed (not in this PR). The new statements optionally support an `as` form:

`preload "my_other_scene.tscn" as other_scene`

So, this doesn't add any new functionality, just syntax. Why should this make a difference?

* users need no longer worry about `const` and `var` when using `load` or `preload`
* allows shorter, less redundant code (gets rid of repetitive `var SOMEBOTHERSOMENAME = ...SOMEBOTHERSOMENAME...` and the impulse to shorten this to `var SBN = ...SOMEBOTHERSOMENAME...`)
* eliminates typing mistakes and bogus naming by consistently deriving the loaded constant/variable name from the loaded path name (the exact rules need to be defined)
* the last point also improves code readability as scripts and resources will now always end up with the same name as the loaded resource, so if you see `DiscreteLineSolver.new()` somewhere in the code, you know exactly what you're dealing with (as opposed to `DLS.new()` as it often happens now)
* visibly separates class global loads and preloads from member var declarations by visually breaking up `var` blocks; grouped blocks of preloads are much cleaner to read now (just like `import` blocks in Python) as they have one aligned row of statements:
	```
	preload "...
	preload "...
	preload "...

	load "...
	load "...

	var ...
	var ...
	```
* I dare say that users from Python, who are familiar with `import`, (and a lot of other language users) will feel much more at home with this

This PR only extends the parser (the new statements just rewrite to the classic `var` and `const` forms internally). The existing functionality of `load` and `preload` is unchanged.

### CODE REVIEW

**Comment by Zylann** [comment]
> preload and load declare suitably named constants and variables derived from the loaded resource's file name

I don't like this. My file structure is `snake_case`, and my class naming is `CamelCase`, I prefer to keep control on my convention and not have to depend on an automatic behavior that mixes file name and class name, declaring identifiers under the hood. This load=var and preload=const duality should be really simple to get, I use it quite often. If you see confused declarations often I think that's because users who wrote them were simply not reasonably experimented yet.

```gdscript
const CircularJump = preload("res://framework/circular_jump.gd")

func _make_jump():
    _jump = CircularJump.new()
```
With auto const:

```gdscript
preload "res://framework/circular_jump.gd"

func _make_jump():
    _jump = CircularJump.new() # Where does this come from?
```

About your Python argument: it is working in Python because files are modules... aka namespaces, not classes. In Godot, script files have the same functionality as classes, hence my problem explained above.
You would also loose control if your file happens to be named the same as a variable in your script or another class.

**Comment by karroffel** [comment]
@Zylann it is mentioned that `as` can be used to name a resource.

```gdscript
preload "res://framework/circular_jump.gd" as CircularJump

func _make_jump():
    _jump = CircularJump.new()
```

**Comment by Zylann** [comment]
I am blind

**Comment by eon-s** [comment]
I don't know, current method makes clear you are loading a Resource, and I never load resources as constants (maybe as exported var), what I get is just a reference, right?

**Comment by vnen** [comment]
@eon-s 
> I don't know, current method makes clear you are loading a Resource,

I really don't see much difference between `var x = preload('x.gd')` and just `preload 'x.gd'`, in terms of clarity that it's a resource.

> and I never load resources as constants (maybe as exported var), what I get is just a reference, right?

I prefer to load as constants. For exports, I rarely have a default resource value, usually I just as `null`. And, of course, the method presented here is not a replacement, but an addition.

**Comment by vnen** [comment]
There's also a couple of general points I want to address:

* File name convention is `snake_case` while class name convention is `PascalCase`. This will inevitably generate statements like `preload "res://my_class.gd" as MyClass`, which doesn't help much to solve redundancy (and I'm not fond of automatic conversion for this case).

----

* How to deal with inner classes? Can one do `preload "res://main_class.gd".InnerClass`? Or something like that?
* Or do you want to go the JS way for more powerful packaging: `preload { InnerClassA, InnerClassB } from "res://util.gd"`? (Don't know how Python manages that)

IMO the package style is much more interesting since it's quite cumbersome to work around, requiring multiple statements to achieve this effect, While I agree that "less typing is not a good argument", in this case it makes clearer that those classes comes from an external file.

**Comment by poke1024** [comment]
@vnen

Importing inner classes to me seems like a bit of a border case. To me, inner classes usually model some internal private aspect of the outer class that is not interesting/visible outside of its embedding outer class. As soon as I need an inner class in a different outer class, I usually would make the inner class a full class in its own file. I personally like your first proposal (which is also along the lines of Java).

The file name/class name convention difference is really a hinderance for this. I'm not sure. Ultimately, to me, it's not really logical to have a class `MyClass` in a file named `my_class.gd`, but since that's the convention I think auto-conversion is the least evil of all the options. This could be turned around though; i.e. `preload MyClass` could just do a `preload my_class.gd` internally.

**Comment by mhilbrunner** [comment]
As discussed in the meeting today, this will be obsoleted by named classes.

**Comment by akien-mga** [comment]
Named classes have now been implemented in the master branch BTW: a3f1ee5c5.
